integration-tests/src/tests/nearcore/sync_state_nodes.rs

```diff
@@ -530,3 +535,141 @@ fn sync_state_dump() {
+                        .process_block_test_no_produce_chunk(
+                            MaybeValidated::from(block.clone()),
+                            Provenance::PRODUCED,
+                        )
+                        .unwrap();
+                    tracing::info!(
+                        "Block {i}: {:?} -- produced no chunk",
+                        block.header().epoch_id()
+                    );
+                } else {
+                    env.process_block(0, block.clone(), Provenance::PRODUCED);
+                    tracing::info!(
+                        "Block {i}: {:?} -- also produced a chunk",
+                        block.header().epoch_id()
+                    );
+                }
+                env.process_block(1, block, Provenance::NONE);
+
+                let tx = SignedTransaction::send_money(
+                    i + 1,
+                    "test0".parse().unwrap(),
+                    "test1".parse().unwrap(),
+                    &signer,
+                    1,
+                    *genesis_block.hash(),
+                );
+                assert_eq!(env.clients[0].process_tx(tx, false, false), ProcessTxResponse::ValidTx);
+            }
+
+            // Simulate state sync
+
+            // No blocks were skipped, therefore we can compute the block height of the first block of the current epoch.
+            let sync_hash_height = ((target_height / epoch_length) * epoch_length + 1) as usize;
+            let sync_hash = *blocks[sync_hash_height].hash();
+            assert_ne!(
+                blocks[sync_hash_height].header().epoch_id(),
+                blocks[sync_hash_height - 1].header().epoch_id()
+            );
+
+            let state_sync_header =
+                env.clients[0].chain.get_state_response_header(0, sync_hash).unwrap();
+            let state_root = state_sync_header.chunk_prev_state_root();
+            let state_root_node = state_sync_header.state_root_node();
+            let num_parts = get_num_state_parts(state_root_node.memory_usage);
+            // Check that state parts can be obtained.
+            let state_parts: Vec<_> = (0..num_parts)
+                .map(|i| {
+                    // This should obviously not fail, aka succeed.
+                    env.clients[0].chain.get_state_response_part(0, i, sync_hash).unwrap()
+                })
+                .collect();
+
+            env.clients[1].chain.reset_data_pre_state_sync(sync_hash).unwrap();
+            let epoch_id = blocks.last().unwrap().header().epoch_id();
+            for i in 0..num_parts {
+                env.clients[1]
+                    .runtime_adapter
+                    .apply_state_part(
+                        0,
+                        &state_root,
+                        PartId::new(i, num_parts),
+                        &state_parts[i as usize],
+                        &epoch_id,
+                    )
+                    .unwrap();
+            }
+        }
+    });
```

Comment:
do we want to also do `set_state_header` and `set_state_finalize` for clients[1] and also check that the data of clients[1] is the same as data of clients[0] after state sync? like `test_catchup_gas_price_change()`